### PR TITLE
WIP: Ethereum oracles

### LINKED ISSUE #881: Specification for Ethereum Oracles
In order to be ready for Q2 2022 development we need to agree upon the scope for the engineering team to deliver in this time period.

We require a spec for Ethereum Oracles

Related to: [Notion Board](https://www.notion.so/vegaprotocol/Flexible-oracle-framework-including-support-for-oracles-on-Ethereum-and-other-bridges-and-basic-comb-8705504dc20c47b3ae581d85f55c3b7b)

How do we know when we are done:

- [ ] The spec has been updated to indicate what are the requirements for flexible oracles and basic combinators
- [ ] The specification contains acceptance criteria with IDs that reflect the scope of the work required for Q2 2022
- [ ] The engineering team has reviewed the specifications and understand the scope
- [ ] The spec is merged into the specs repo before engineering start development/test work

### PR DESCRIPTION
**Pull request by barnabee**
close #881 .

### CODE REVIEW

File: protocol/0000-ETHD-ethereum-data-source.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
+# Ethereum data source
+
+
+## Summary
+
+This specification adds a new way of sourcing data from the Ethereum blockchain to allow arbitrary data from the Ethereum blockchain to be ingested as a [data source](./0045-DSRC-data_sourcing.md).
```
Comment (AnExsomnis) [inline protocol/0000-ETHD-ethereum-data-source.md]:
```suggestion
This specification adds a new way of sourcing data from the Ethereum blockchain to allow arbitrary data from the Ethereum blockchain to be ingested as a [data source](./0045-DSRC-data_sourcing.md) into the Vega network.
```

File: protocol/0000-ETHD-ethereum-data-source.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+The event queue may evaluate filters before submitting the observation to the Vega chain, in order to avoid submitting transactions containing data that will be dropped due to filters in any case.
+
+
+## Functional design
+
+### Contracts and ABI
+
+An Ethereum oracle has as its subject a smart contract that is deployed on the Ethereum blockchain.
+All such contracts have an address, and an "ABI" that defines the methods an events exposed by the contract.
+In order interpret the oracle specification and interact with the smart contract, both the contract address and Ethereum ABI JSON for the contract (or a subset, covering the relevant parts) must therefore be included in the oracle specification.
```
Comment (AnExsomnis) [inline protocol/0000-ETHD-ethereum-data-source.md]:
```suggestion
In order to interpret the oracle specification and interact with the smart contract, both the contract address and Ethereum ABI JSON for the contract (or a subset, covering the relevant parts) must therefore be included in the oracle specification.
```

File: protocol/0000-ETHD-ethereum-data-source.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+Event data and data returned from functions will be emitted by the Ethereum node in an ABI-encoded format.
+This data, including any structs should be decoded using the ABI into a JSON-like representation.
+
+Note: as with any data source containing JSON formatted (or other arbitrary structured) data, the data required by the consumer of the data source may be a fields or sub-objects (including nested fields and objects).
```
Comment (AnExsomnis) [inline protocol/0000-ETHD-ethereum-data-source.md]:
```suggestion
Note: as with any data source containing JSON formatted (or other arbitrary structured) data, the data required by the consumer of the data source may consist of a number of fields or sub-objects (including nested fields and objects).
```

File: protocol/0000-ETHD-ethereum-data-source.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+### Ethereum chain data enrichment
+
+All data sourced from Ethereum should be structured as an object containing both a payload and Ethereum chain metadata, namely:
+
+- Ethereum block height at which the data was observed/event occurred
+
+- Ethereum block timestamp at which the data was observed/event occurred
+
+
+These data can be used as the subject of filters or even extracted as the oracle data of interest.
```
Comment (AnExsomnis) [inline protocol/0000-ETHD-ethereum-data-source.md]:
```suggestion
This data can be used as the subject of filters or even extracted as the oracle data of interest.
```

File: protocol/0000-ETHD-ethereum-data-source.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+1. Create ethereum oracles based on calling a read method of a smart contract (Phase 2 - oracle based on listening for events)
+2. All current governance rules that apply to propose / submit / vote on a proposal should be applicable for the ethereum oracle data source creation / amendment
+3. Create more than spam.protection.max.proposals oracle data source proposals in an epoch - proposal rejected with error message
+4. Create ethereum oracles based on calling a read method of a smart contract by supplying incorrect ABI (Phase 2 - oracle based on listening for events)
+
+### External Oracles - Amendments
+
+1. Amend the oracle data source via governance proposals. Amendments should take effect as soon as the proposal is enacted.
+2. Amend an existing ethereum data source and change the contract address and enact the proposal. Once enacted , the data should be sourced from the new smart contract. Try amending the other fields in the oracle data source and the changes should take effect after the proposals are enacted.
```
Comment (AnExsomnis) [inline protocol/0000-ETHD-ethereum-data-source.md]:
```suggestion
2. Amend an existing ethereum data source and change the contract address and enact the proposal. Once enacted, the data should be sourced from the new smart contract. Try amending the other fields in the oracle data source and the changes should take effect after the proposals are enacted.
```

**Review by ettec — DISMISSED** [review]
Fine for now

File: protocol/0081-ETHD-ethereum-data-source.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+The event queue may evaluate filters before submitting the observation to the Vega chain, in order to avoid submitting transactions containing data that will be dropped due to filters in any case.
+
+
+## Functional design
+
+### Contracts and ABI
+
+An Ethereum oracle has as its subject a smart contract that is deployed on the Ethereum blockchain.
+All such contracts have an address, and an "ABI" that defines the methods an events exposed by the contract.
+In order interpret the oracle specification and interact with the smart contract, both the contract address and Ethereum ABI JSON for the contract (or a subset, covering the relevant parts) must therefore be included in the oracle specification.
```
Comment (AnExsomnis) [inline protocol/0081-ETHD-ethereum-data-source.md]:
```suggestion
In order to interpret the oracle specification and interact with the smart contract, both the contract address and Ethereum ABI JSON for the contract (or a subset, covering the relevant parts) must therefore be included in the oracle specification.
```

File: protocol/0081-ETHD-ethereum-data-source.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+Event data and data returned from functions will be emitted by the Ethereum node in an ABI-encoded format.
+This data, including any structs should be decoded using the ABI into a JSON-like representation.
+
+Note: as with any data source containing JSON formatted (or other arbitrary structured) data, the data required by the consumer of the data source may be a fields or sub-objects (including nested fields and objects).
```
Comment (AnExsomnis) [inline protocol/0081-ETHD-ethereum-data-source.md]:
```suggestion
Note: as with any data source containing JSON formatted (or other arbitrary structured) data, the data required by the consumer of the data source may be a number of fields or sub-objects (including nested fields and objects).
```

File: protocol/0081-ETHD-ethereum-data-source.md [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+### Ethereum chain data enrichment
+
+All data sourced from Ethereum should be structured as an object containing both a payload and Ethereum chain metadata, namely:
+
+- Ethereum block height at which the data was observed/event occurred
+
+- Ethereum block timestamp at which the data was observed/event occurred
+
+
+These data can be used as the subject of filters or even extracted as the oracle data of interest.
```
Comment (AnExsomnis) [inline protocol/0081-ETHD-ethereum-data-source.md]:
```suggestion
This data can be used as the subject of filters or even extracted as the oracle data of interest.
```

File: protocol/0081-ETHD-ethereum-data-source.md [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+3. Create more than spam.protection.max.proposals oracle data source proposals in an epoch - proposal rejected with error message (<a name="0081-ETHD-003" href="#0081-ETHD-003">0081-ETHD-003</a>)
+4. Create ethereum oracles based on calling a read method of a smart contract by supplying incorrect ABI (Phase 2 - oracle based on listening for events) (<a name="0081-ETHD-004" href="#0081-ETHD-004">0081-ETHD-004</a>)
+
+### External Oracles - Amendments
+
+1. Amend the oracle data source via governance proposals. Amendments should take effect as soon as the proposal is enacted (<a name="0081-ETHD-005" href="#0081-ETHD-005">0081-ETHD-005</a>)
+2. Amend an existing ethereum data source and change the contract address and enact the proposal. Once enacted , the data should be sourced from the new smart contract. Try amending the other fields in the oracle data source and the changes should take effect after the proposals are enacted (<a name="0081-ETHD-006" href="#0081-ETHD-006">0081-ETHD-006</a>)
```
Comment (AnExsomnis) [inline protocol/0081-ETHD-ethereum-data-source.md]:
```suggestion
2. Amend an existing ethereum data source and change the contract address and enact the proposal. Once enacted, the data should be sourced from the new smart contract. Try amending the other fields in the oracle data source and the changes should take effect after the proposals are enacted (<a name="0081-ETHD-006" href="#0081-ETHD-006">0081-ETHD-006</a>)
```

**Review by AnExsomnis — CHANGES_REQUESTED** [review]

**Review by AnExsomnis — DISMISSED** [review]

**Review by AnExsomnis — DISMISSED** [review]

File: protocol/0082-ETHD-ethereum-data-source.md [hunk 10]
```diff
@@ -110,84 +110,84 @@ Select {
...
-2. All current governance rules that apply to propose / submit / vote on a proposal should be applicable for the ethereum oracle data source creation / amendment (<a name="0081-ETHD-002" href="#0081-ETHD-002">0081-ETHD-002</a>)
-3. Create more than spam.protection.max.proposals oracle data source proposals in an epoch - proposal rejected with error message (<a name="0081-ETHD-003" href="#0081-ETHD-003">0081-ETHD-003</a>)
-4. Create ethereum oracles based on calling a read method of a smart contract by supplying incorrect ABI (Phase 2 - oracle based on listening for events) (<a name="0081-ETHD-004" href="#0081-ETHD-004">0081-ETHD-004</a>)
+1. Create ethereum oracles based on calling a read method of a smart contract (Phase 2 - oracle based on listening for events) (<a name="0082-ETHD-001" href="#0082-ETHD-001">0082-ETHD-001</a>)
+2. All current governance rules that apply to propose / submit / vote on a proposal should be applicable for the ethereum oracle data source creation / amendment (<a name="0082-ETHD-002" href="#0082-ETHD-002">0082-ETHD-002</a>)
```
Comment (barnabee) [inline protocol/0082-ETHD-ethereum-data-source.md]: Is there a specific data source creation proposal or is this talking about proposals such as market creation/amendment that introduce a new Eth oracle?

File: protocol/0082-ETHD-ethereum-data-source.md [hunk 11]
```diff
@@ -110,84 +110,84 @@ Select {
...
+3. Phase 2 - Amend an existing ethereum data source and change the events that we are listening to and enact the proposal. Once enacted , the data should be sourced from the amended events. Try amending the other fields in the oracle data source and the changes should take effect after the proposals are enacted (<a name="0082-ETHD-007" href="#0082-ETHD-007">0082-ETHD-007</a>)
 
 ### External Oracles - Deletions
 
-1. Ability to delete a data source if and only if the data source is NOT used by any active markets (<a name="0081-ETHD-008" href="#0081-ETHD-008">0081-ETHD-008</a>)
-2. Should NOT be able to delete a data source if its being actively used by a market (<a name="0081-ETHD-009" href="#0081-ETHD-009">0081-ETHD-009</a>)
-3. If a single data source is used by multiple markets, then should NOT be able to delete the data source even if one of those markets is actively using the data source (<a name="0081-ETHD-010" href="#0081-ETHD-010">0081-ETHD-010</a>)
+1. Ability to delete a data source if and only if the data source is NOT used by any active markets (<a name="0082-ETHD-008" href="#0082-ETHD-008">0082-ETHD-008</a>)
```
Comment (barnabee) [inline protocol/0082-ETHD-ethereum-data-source.md]: Can you explain the intended mechanics for creating/deleting data sources if it does not happen automatically when they are first reference (create) and no longer reference (delete)? If it is automatic I think the AC could be made clearer.

File: protocol/0082-ETHD-ethereum-data-source.md [hunk 12]
```diff
@@ -110,84 +110,84 @@ Select {
...
-4. Create a new market with an inactive external oracle data source, system should throw an error (<a name="0081-ETHD-014" href="#0081-ETHD-014">0081-ETHD-014</a>)
-5. Validations to be applied - need to be expanded (<a name="0081-ETHD-015" href="#0081-ETHD-015">0081-ETHD-015</a>)
-6. Any mismatch between expected fields and received fields should emit an error via the TX RESULT event (<a name="0081-ETHD-016" href="#0081-ETHD-016">0081-ETHD-016</a>)
+1. Validate if the smart contract address is valid (<a name="0082-ETHD-011" href="#0082-ETHD-011">0082-ETHD-011</a>)
+2. Validate if the data elements of the oracle data source is valid - e.g. source for a value that's returned as boolean but have a filter / condition for greater than 0 (<a name="0082-ETHD-012" href="#0082-ETHD-012">0082-ETHD-012</a>)
+3. Validations for min / max frequency of listening for events / read a smart contract (<a name="0082-ETHD-013" href="#0082-ETHD-013">0082-ETHD-013</a>)
+4. Create a new market with an inactive external oracle data source, system should throw an error (<a name="0082-ETHD-014" href="#0082-ETHD-014">0082-ETHD-014</a>)
```
Comment (barnabee) [inline protocol/0082-ETHD-ethereum-data-source.md]: What does this mean? If the contract address, ABI, etc. are valid why would it error?

File: protocol/0082-ETHD-ethereum-data-source.md [hunk 13]
```diff
@@ -110,84 +110,84 @@ Select {
...
+5. Validations to be applied - need to be expanded (<a name="0082-ETHD-015" href="#0082-ETHD-015">0082-ETHD-015</a>)
```
Comment (barnabee) [inline protocol/0082-ETHD-ethereum-data-source.md]: Is this an unfinished AC?

File: protocol/0082-ETHD-ethereum-data-source.md [hunk 14]
```diff
@@ -110,84 +110,84 @@ Select {
...
+6. Any mismatch between expected fields and received fields should emit an error via the TX RESULT event (<a name="0082-ETHD-016" href="#0082-ETHD-016">0082-ETHD-016</a>)
 
 ### New Network parameters
 
-1. Test min / max values / validations for any new network parameters that are added (<a name="0081-ETHD-017" href="#0081-ETHD-017">0081-ETHD-017</a>)
-2. Test the impact / behaviour of the system, after the changes to the new network parameters are enacted (<a name="0081-ETHD-018" href="#0081-ETHD-018">0081-ETHD-018</a>)
+1. Test min / max values / validations for any new network parameters that are added (<a name="0082-ETHD-017" href="#0082-ETHD-017">0082-ETHD-017</a>)
+2. Test the impact / behaviour of the system, after the changes to the new network parameters are enacted (<a name="0082-ETHD-018" href="#0082-ETHD-018">0082-ETHD-018</a>)
```
Comment (barnabee) [inline protocol/0082-ETHD-ethereum-data-source.md]: I assume this (and probably the previous AC) would need to be expanded to multiple ACs?

File: protocol/0082-ETHD-ethereum-data-source.md [hunk 15]
```diff
@@ -110,84 +110,84 @@ Select {
...
+3. Phase 2 - Define behaviour for missed events / missed scheduled smart contract calls - e.g. if an oracle data source is scheduled to emit events every 10 minutes and we miss 5 events because of protocol upgrade or some other outage - then do we catch up those events or skip those events ? Maybe this is defined in the oracle data source definition (<a name="0082-ETHD-029" href="#0082-ETHD-029">0082-ETHD-029</a>)
+4. If an oracle data source is inactive - then any events / any data received from that oracle data source is NOT processed (<a name="0082-ETHD-030" href="#0082-ETHD-030">0082-ETHD-030</a>)
+5. SPAM rules if any defined should be tested for (<a name="0082-ETHD-031" href="#0082-ETHD-031">0082-ETHD-031</a>)
+6. NOT all data sourced should be stored on chain - invalid / incorrect data is filtered out and is NOT processed / stored on chain - understand what the rules are and design the AC's / test accordingly (<a name="0082-ETHD-032" href="#0082-ETHD-032">0082-ETHD-032</a>)
+7. Any active data sources that aren't used by any markets should not source data until they are being actively used by a market (<a name="0082-ETHD-033" href="#0082-ETHD-033">0082-ETHD-033</a>)
```
Comment (barnabee) [inline protocol/0082-ETHD-ethereum-data-source.md]: What is the difference between "active" and "actively [used by a market/referenced by something within Vega core]? Why does this difference exist?

File: protocol/0082-ETHD-ethereum-data-source.md [hunk 16]
```diff
@@ -110,84 +110,84 @@ Select {
...
 
 ### Usage
 
-1. It should be possible to use only ethereum oracle data sources or internal data sources or use a combination of both types of oracles (<a name="0081-ETHD-034" href="#0081-ETHD-034">0081-ETHD-034</a>)
-2. Create a market to use an internal data source to terminate a market and an ethereum oracle to settle the market (<a name="0081-ETHD-035" href="#0081-ETHD-035">0081-ETHD-035</a>)
-3. Create a market to use an external data source to terminate a market and an internal / manual oracle to settle the market (<a name="0081-ETHD-036" href="#0081-ETHD-036">0081-ETHD-036</a>)
-4. Data sourcing should be completely decoupled from data filtering (<a name="0081-ETHD-037" href="#0081-ETHD-037">0081-ETHD-037</a>)
+1. It should be possible to use only ethereum oracle data sources or internal data sources or use a combination of both types of oracles (<a name="0082-ETHD-034" href="#0082-ETHD-034">0082-ETHD-034</a>)
```
Comment (barnabee) [inline protocol/0082-ETHD-ethereum-data-source.md]: Clarify. I assume this means e.g. using one type for trading termination and the other for settlement, the equivalent for perps?

**Review by barnabee — COMMENTED** [review]
Some questions/comments on the changes to this. I can only "Comment" not "Request changes" so I've done that.

Would appreciate some guidance @gordsport on how finished the ACs should be. They seem high level/unfinished in places.

Also the changes/ACs appear to suggest additional or different functionality to that implied and required purely by the spec and existing [overriding data sourcing principles](https://github.com/vegaprotocol/specs/blob/master/protocol/0045-DSRC-data_sourcing.md) that are currently expected to apply to all data sources, for instance around tracking active data sources.

File: protocol/0082-ETHD-ethereum-data-source.md (hunk 10 above)
Comment (AnExsomnis) [reply protocol/0082-ETHD-ethereum-data-source.md]: Yes, the governance path that creates/updates a market is the intended way to handle this kind of data source, in the same way as the other ones.

File: protocol/0082-ETHD-ethereum-data-source.md (hunk 11 above)
Comment (AnExsomnis) [reply protocol/0082-ETHD-ethereum-data-source.md]: The intended mechanism to deactivate and delete a data source is the one we have at the moment.
Seems a little rewording will help, I'll suggest changes.

File: protocol/0082-ETHD-ethereum-data-source.md (hunk 15 above)
Comment (AnExsomnis) [reply protocol/0082-ETHD-ethereum-data-source.md]: I think this point should be removed, there is no such difference that it is implying

**Comment by AnExsomnis** [comment]
Hello @barnabee , I addressed your comments and did some rewording in this PR https://github.com/vegaprotocol/specs/pull/1834
It is opened from this branch.

**Review by AnExsomnis — DISMISSED** [review]

**Review by AnExsomnis — APPROVED** [review]

**Pull request merged**